src/common/marlin_server.cpp

```diff
@@ -1557,9 +1555,9 @@ void fsm_destroy(ClientFSM type) {
 
 void _fsm_change(ClientFSM type, fsm::BaseData data) {
     {
-        static int previous_type = -1;
-        static int previous_phase = -1;
-        if (previous_type != static_cast<int>(type) || previous_phase != static_cast<int>(data.GetPhase())) {
+        static std::optional<int> previous_type;
+        static std::optional<int> previous_phase;
+        if (previous_type.value_or(-1) != static_cast<int>(type) || previous_phase.value_or(-1) != static_cast<int>(data.GetPhase())) {
```

Comment:
The value_or is useful if you actually have a fallback value. That -1 doesn't really make sense here.

Nevertheless, if your only intention is to make sure `nullopt` is different than the value, then I believe this Just Works ‒ `previous_type != static_cast(...)` will make it different if it's not there.